### PR TITLE
Feature/zcs 11179

### PR DESCRIPTION
**Pull request by SuryawanshiAmol**
- The core, event, and http directory code has been merged from branches/stable 1.24 to the current zimbra/nginx.

### CODE REVIEW

**Comment by ghen2** [comment]
Btw minor build warning:
`./auto/configure: warning: the "--with-ipv6" option is deprecated`
(since nginx 1.11.5)

**Comment by ghost** [comment]
When comparing our changes from the 1.20 version next to our changes in the 1.24 version, it seems like we omit some of our customizations in ngx_mail_parse.c beginning under `/* IMAP tag */` we used to add `case '\x0'` but we do not anymore. Can you explain why we make a different approach now?


ngx_mail_parse c

**Comment by SuryawanshiAmol** [comment]
When comparing our changes from the 1.20 version next to our changes in the 1.24 version, it seems like we omit some of our customizations in ngx_mail_parse.c beginning under /* IMAP tag */ we used to add case '\x0' but we do not anymore. Can you explain why we make a different approach now?
Answer: I've left it out because it's now handled in the default case. See the code below the '\x0' will satisfy the if condition.
    u_char ch('\x0');
    if ((ch < 'A' || ch > 'Z') && (ch < 'a' || ch > 'z')
        && (ch < '0' || ch > '9') && ch != '-' && ch != '.'
        && ch != '_')
    {
        cout << "goto Invalid";
    }

Or we can handle it as below:
case CR:
case LF:
**case '\x0':** 
       goto invalid;

**Comment by SuryawanshiAmol** [comment]
The --with-ipv6 option no longer needed. But no harm keeping it. Its not added recently its their since beginning.
Please See: 
https://nginx.org/en/CHANGES-1.20

Changes with nginx 1.11.5                                        11 Oct 2016

    *) Change: the --with-ipv6 configure option was removed, **now IPv6
       support is configured automatically.**

**Comment by ghost** [comment]
In ngx_mail_parse.c the \x0 is also missing swi_spaces_before_command method. If that is fixed, the rest of the customization in this file look the same as in the 1.20 version.

**Review by ghost — APPROVED** [review]

File: src/event/ngx_event_openssl.c [hunk 1]
```diff
@@ -2896,25 +3198,27 @@ ngx_int_t
 ngx_ssl_shutdown(ngx_connection_t *c)
 {
     int         n, sslerr, mode;
+    ngx_int_t   rc;
     ngx_err_t   err;
     ngx_uint_t  tries;
 
+    rc = NGX_OK;
+// Zimbra customizations start here (Jira Tickets: )	
+//	ngx_ssl_ocsp_cleanup(c);
```
Comment (desouzas) [inline src/event/ngx_event_openssl.c]: Looks like this is done below the if statement (as customized before this update), can remove.

File: src/mail/ngx_mail_parse.c [hunk 2]
```diff
@@ -716,32 +758,44 @@ ngx_mail_imap_parse_command(ngx_mail_session_t *s)
 done:
 
     s->buffer->pos = p + 1;
-
```
Comment (desouzas) [inline src/mail/ngx_mail_parse.c:719]: Is this not needed: https://github.com/nginx/nginx/blob/release-1.24.0/src/mail/ngx_mail_parse.c#L636
`s->state = (s->command != NGX_IMAP_AUTHENTICATE) ? swi_start : swi_argument;`

File: src/mail/ngx_mail_parse.c [hunk 3]
```diff
@@ -1427,36 +1479,26 @@ ngx_mail_smtp_parse_command(ngx_mail_session_t *s)
 
     s->buffer->pos = p + 1;
```
Comment (desouzas) [inline src/mail/ngx_mail_parse.c:1481]: Is this not needed: https://github.com/nginx/nginx/blob/release-1.24.0/src/mail/ngx_mail_parse.c#L879
`s->state = (s->command != NGX_SMTP_AUTH) ? sws_start : sws_argument;`

File: auto/configure [hunk 4]
```diff
@@ -44,6 +44,7 @@ if test -z "$NGX_PLATFORM"; then
 else
     echo "building for $NGX_PLATFORM"
     NGX_SYSTEM=$NGX_PLATFORM
+    NGX_MACHINE=i386
```
Comment (RakeshAMore) [inline auto/configure:47]: Do we want to hardcode this to intel now?

File: auto/options [hunk 5]
```diff
@@ -582,10 +592,12 @@ cat << END
   --with-pcre=DIR                    set path to PCRE library sources
   --with-pcre-opt=OPTIONS            set additional build options for PCRE
   --with-pcre-jit                    build PCRE with JIT compilation support
+  --without-pcre2                    do not use PCRE2 library
```
Comment (RakeshAMore) [inline auto/options:595]: @umagmrit I think you should have attention here.

File: src/event/ngx_event_openssl.c (hunk 1 above)
Comment (SuryawanshiAmol) [reply src/event/ngx_event_openssl.c]: Correct. Removed.

File: src/mail/ngx_mail_parse.c (hunk 2 above)
Comment (SuryawanshiAmol) [reply src/mail/ngx_mail_parse.c:719]: No It is not required. We carefully eliminated it.
Same can be referenced here -> (when we moved from 1.19.0 to 1.20.0 https://github.com/Zimbra/nginx/commit/3c9e50fdeb46693d5bbc53028824dccc488fbdeb#diff-f2fa0c604b3eec568bf77394733618361e7ad69d89836b129ff05cfea2d12244

File: src/mail/ngx_mail_parse.c (hunk 3 above)
Comment (SuryawanshiAmol) [reply src/mail/ngx_mail_parse.c:1481]: No It is not required. We carefully eliminated it.
Same can be referenced here -> (when we moved from 1.19.0 to 1.20.0 https://github.com/Zimbra/nginx/commit/3c9e50fdeb46693d5bbc53028824dccc488fbdeb#diff-f2fa0c604b3eec568bf77394733618361e7ad69d89836b129ff05cfea2d12244

File: auto/configure (hunk 4 above)
Comment (SuryawanshiAmol) [reply auto/configure:47]: It appears to be a fix for cross-platform os/platform construction. It's not our customization; it's an NGINX team fix. 
Please see commit: https://github.com/nginx/nginx/commit/62b790c3318a56971e461a1e1f5b79197b604e89

File: auto/options (hunk 5 above)
Comment (SuryawanshiAmol) [reply auto/options:595]: Its looks to be some sort of fix for disabling the PCRE2 library. Its not our customization its fix from NGINX team in public repo. 
Please see: https://github.com/nginx/nginx/commit/c6fec0b027569a4e0b1d8aaee7dea0f2e4d6052b

**Comment by ghen2** [comment]
Hi

Could you consider including https://github.com/nginx/nginx/commit/6ceef192e7af1c507826ac38a2d43f08bf265fb9 to mitigate [CVE-2023-44487: HTTP/2 Rapid Reset Attack](https://thehackernews.com/2023/10/http2-rapid-reset-zero-day.html)?  This patch will be part of an upcoming nginx 1.24.x release.

It applies cleanly to nginx 1.20.0 as well.

**Comment by ghen2** [comment]
I created PR https://github.com/Zimbra/nginx/pull/7 for Zimbra's nginx 1.20.0, we're already running with this patch in production.

**Pull request merged**